File: src/highs_lp.cpp

#include <chrono>
#include <cstdio>
#include <string>
#include <vector>

#include <Highs.h>

#include "graph_io.hpp"

int main(int argc, char* argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: highs_lp <graph_file> [--time-limit SECONDS]\n");
        return 1;
    }
    double time_limit = parse_time_limit(argc, argv, 300.0);

    Graph g;
    try {
        g = read_graph(argv[1]);
    } catch (const std::exception& e) {
        print_result(-1, 0.0, "error", e.what());
        return 1;
    }
    HighsInt m = (HighsInt)g.edges.size();

    Highs h;
    h.setOptionValue("output_flag", false);
    h.setOptionValue("solver", "simplex");
    h.setOptionValue("simplex_strategy", 4);
    h.setOptionValue("time_limit", time_limit);
    h.changeObjectiveSense(ObjSense::kMaximize);

    std::vector<double> lbs(m, 0.0), ubs(m, 1.0);
    h.addVars(m, lbs.data(), ubs.data());
    if (m > 0) {
        std::vector<double> costs(m, 1.0);
        h.changeColsCost(0, m - 1, costs.data());
    }

    std::vector<std::vector<HighsInt>> left_cols(g.n_left), right_cols(g.n_right);
    for (HighsInt i = 0; i < m; ++i) {
        left_cols[g.edges[i].first].push_back(i);
        right_cols[g.edges[i].second].push_back(i);
    }

    std::vector<double>   row_lb, row_ub;
    std::vector<HighsInt> starts, indices;
    std::vector<double>   values;
    HighsInt nz = 0;
    auto add_row = [&](const std::vector<HighsInt>& cols) {
        row_lb.push_back(-1e30);
        row_ub.push_back(1.0);
        starts.push_back(nz);
        for (HighsInt c : cols) { indices.push_back(c); values.push_back(1.0); ++nz; }
    };
    for (int u = 0; u < g.n_left;  ++u) if (!left_cols[u].empty())  add_row(left_cols[u]);
    for (int w = 0; w < g.n_right; ++w) if (!right_cols[w].empty()) add_row(right_cols[w]);

    HighsInt num_rows = (HighsInt)starts.size();
    if (num_rows > 0) {
        h.addRows(num_rows, row_lb.data(), row_ub.data(),
                  nz, starts.data(), indices.data(), values.data());
    }

    auto t0 = std::chrono::steady_clock::now();
    HighsStatus run_status = h.run();
    auto t1 = std::chrono::steady_clock::now();
    double elapsed = std::chrono::duration<double>(t1 - t0).count();

    HighsModelStatus model_status = h.getModelStatus();

    if (model_status == HighsModelStatus::kTimeLimit) {
        print_result(-1, elapsed, "timeout", "HiGHS hit time limit");
        return 0;
    }
    if (model_status != HighsModelStatus::kOptimal) {
        char buf[120];
        snprintf(buf, sizeof(buf),
                 "HiGHS non-optimal status (model=%d, run=%d)",
                 (int)model_status, (int)run_status);
        print_result(-1, elapsed, "error", buf);
        return 1;
    }

    double objval = h.getObjectiveValue();
    print_result((int)(objval + 0.5), elapsed, "optimal");
    return 0;
}

File: python/solvers/highs_lp.py


import time

from ._common import make_result, memory_monitor, read_graph


def solve(graph_path: str, time_limit: float = 300.0) -> dict:
    try:
        import highspy
    except ImportError as e:
        return make_result(status="error", error_message=f"highspy import failed: {e}")

    with memory_monitor() as peak_mb:
        t0 = time.perf_counter()
        try:
            n_left, n_right, edges = read_graph(graph_path)
            m = len(edges)

            if m == 0:
                return make_result(
                    matching_size=0,
                    time_seconds=time.perf_counter() - t0,
                    peak_memory_mb=peak_mb[0],
                    status="optimal",
                )

            lp = highspy.HighsLp()
            lp.num_col_ = m
            lp.num_row_ = 0
            lp.col_cost_ = [1.0]*m
            lp.col_lower_ = [0.0]*m
            lp.col_upper_ = [1.0]*m
            lp.sense_ = highspy.ObjSense.kMaximize

            left_edge = [[] for _ in range(n_left)]
            right_edge = [[] for _ in range(n_right)]
            for idx, (u, w) in enumerate(edges):
                left_edge[u].append(idx)
                right_edge[w].append(idx)

            row_lower, row_upper, starts, indices, vals = [], [], [], [], []
            nz = 0
            for u in range(n_left):
                if left_edge[u]:
                    row_lower.append(-1e30)
                    row_upper.append(1.0)
                    starts.append(nz)
                    for c in left_edge[u]:
                        indices.append(c); vals.append(1.0)
                    nz += len(left_edge[u])
            for w in range(n_right):
                if right_edge[w]:
                    row_lower.append(-1e30)
                    row_upper.append(1.0)
                    starts.append(nz)
                    for c in right_edge[w]:
                        indices.append(c); vals.append(1.0)
                    nz += len(right_edge[w])

            num_rows = len(row_lower)
            lp.num_row_ = num_rows
            lp.row_lower_ = row_lower
            lp.row_upper_ = row_upper
            lp.a_matrix_.format_ = highspy.MatrixFormat.kRowwise
            lp.a_matrix_.num_col_ = m
            lp.a_matrix_.num_row_ = num_rows
            lp.a_matrix_.start_ = starts + [nz]
            lp.a_matrix_.index_ = indices
            lp.a_matrix_.value_ = vals

            h = highspy.Highs()
            h.setOptionValue("output_flag", False)
            h.setOptionValue("solver", "simplex")
            h.setOptionValue("simplex_strategy", 4)        
            h.setOptionValue("time_limit", float(time_limit))
            h.passModel(lp)
            h.run()

            elapsed = time.perf_counter() - t0
            model_status = str(h.getModelStatus())

            if "TimeLimit" in model_status or "kTimeLimit" in model_status:
                return make_result(
                    time_seconds=elapsed, peak_memory_mb=peak_mb[0],
                    status="timeout",
                    error_message=f"HiGHS hit time limit (status {model_status})",
                )
            if "Optimal" not in model_status and "kOptimal" not in model_status:
                return make_result(
                    time_seconds=elapsed, peak_memory_mb=peak_mb[0],
                    status="error",
                    error_message=f"HiGHS non-optimal status {model_status}",
                )

            sol = h.getSolution()
            col_values = list(sol.col_value)
            for i, v in enumerate(col_values):
                if abs(v - round(v)) > 1e-6:
                    return make_result(
                        time_seconds=elapsed, peak_memory_mb=peak_mb[0],
                        status="error",
                        error_message=f"Non-integral LP value {v} at edge {i}",
                    )

            return make_result(
                matching_size=int(round(sum(col_values))),
                time_seconds=elapsed, peak_memory_mb=peak_mb[0],
                status="optimal",
            )
        except Exception as e:
            return make_result(
                time_seconds=time.perf_counter() - t0,
                peak_memory_mb=peak_mb[0],
                status="error",
                error_message=str(e),
            )
